documentation/general/program-scale-testing-requirements/hero-implementation-details.md

```diff
@@ -0,0 +1,247 @@
+
+
+### "1 IP" test
+
+Minimum config posible to run traffic through. 
+
+#### Objectives
+
+##### validate the hardware and software. 
+
+It ensures we can program the DPU via private API, SAI or DASH and that we can pass 1 packets end to end from traffic generator through the device under test and back.
+
+##### can also provide best case scenario performance numbers
+
+its a maybe because 1 packets replicated milions of times may not necesarly work best for all hardware implementations.
+
+
+### Baby Hero test
+
+It is ment to be an intermediary step between 1 IP test and Hero test. 
+
+We try to keep the scale at ~1% of hero test by using only one prefix per ACL instead of 100
+
+| Metric                       | Baby Hero | % of HERO  |
+|------------------------------|-----------|------------|
+| VNIs                         | 32        | 3%         |
+| ENIs                         | 32        | 100%       |
+| NSGs per ENI                 | 10        | 100%       |
+| NSGs                         | 320       | 100%       |
+| ACL rules per NSG            |  1,000    | 100%       |
+| ACL rules                    |  320,000  | 100%       |
+| ACL prefixes per ACL rule    | 1         | 1%         |
+| ACL prefixes                 |  320,000  | 1%         |
+| ACL ports                    | NA        | 0%         |
+| VNETs (Outbound)             | 32        | 100%       |
+| Outbound routes per ENI      | 5K        |            |
+| Outbound routes              | 160K      | 5%         |
+| CA-to-PA mapping (Outbound)  | 80320     | 1%         |
+| Inbound Routes per ENI       | 1         | 0%         |
+| Inbound Routes               | 32        | 100%       |
+| PAs (Inbound PA Validation)  | 32        | 100%       |
+
+
+### In between scale
+
+If Hero test scale numbers cannot be met we can add another checkpoint to gather aditional data before final implementation is ready
+
+### Hero test
+
+The test exactly as specified in the DASH requirements.
+
+### Best case scenario
+
+If any of the previous tests have not shown the best case scenario we can run a test with the best case scenario in mind.
+
+### Worst case scenario
+
+If any of the previous tests have not shown the worst case scenario we add this test as well (without exceeding hero test scale).
+
+
+## Metrics
+
+Besides CPS we actualy colect for all those dieferent scale points multiple metrics.
+
+### PPS
+
+PPS (packets per second) is the first metric we colect. Usualy it is UDP packets are used
+Packets must be as small as posible so we do not to hit link speed constranints.
+This test shows the fast path perfromance.
+
+### Latency
+
+Latency is the time it takes for a packet to go through the device under test.
+
+Latency value is most acurate when we have highest PPS, smallest packet, and zero packet loss. and is measured using Ixnetwork and Novus card.
+
+Aim for DPU is 2us, for smart switch we have to consider that packet travels twice through the NPU as well.
```

Comment:
Latency through a switch is very dependent upon the queuing and congestion. Are you trying to find minimum latency through the switch? Do you have a way to measure just the switch latency w/o DPUs?

Reply:
yes we can find the NPU/switch latency, it is considered a known variable. since the NPU is usually a 32x400G asic and we use only 8x400G for the test  the NPU is **usually** not a bottleneck or point of issues.

rephrased

"When testing the smart switch we have to run a test to get the switch latency without running the traffic through the DPU and then get the total system latency with the understanding that each packets travel once through the NPU to reach the DPU, than it travel through teh DPU and once more it will travel through teh NPU after it leave the DPU. 

smart switch latency = 2 x NPU latency + DPU latency"